Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs

```diff
@@ -0,0 +1,210 @@
+﻿using System;
+using System.Composition.Hosting.Core;
+using System.Linq;
+using System.Net.Security;
+using Content.Server.GameObjects.Components.Atmos;
+using Content.Server.GameObjects.Components.Body.Circulatory;
+using Content.Server.GameObjects.Components.Body.Respiratory;
+using Content.Server.Utility;
+using Content.Shared.Chemistry;
+using Content.Shared.Interfaces.GameObjects.Components;
+using Robust.Server.GameObjects;
+using Robust.Shared.GameObjects;
+using Robust.Shared.GameObjects.Components.Timers;
+using Robust.Shared.GameObjects.Components.Transform;
+using Robust.Shared.Interfaces.GameObjects;
+using Robust.Shared.Interfaces.Map;
+using Robust.Shared.IoC;
+using Robust.Shared.Maths;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization;
+
+namespace Content.Server.GameObjects.Components.Chemistry
+{
+    [RegisterComponent]
+    public class SmokeComponent : Component
+    {
+        private const float ExposeDelay = 0.5f;
+
+        public override string Name => "Smoke";
+
+        [Dependency] private readonly IMapManager _mapManager = default!;
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
```

Comment:
If you moved the reaction stuff onto the system you could avoid having to store these references on every smoke component.